Add column_accumulate and generalize column_reduce

dennisYatunin opened the pull request:
<!-- Provide a clear description of the content -->

This PR adds a `column_accumulate!` function that can `accumulate` values along each column of a `Field`, a pointwise `Broadcasted` object, or a columnwise `StencilBroadcasted` object. It also generalizes `column_mapreduce!`, which could only accept `Field`s, to `column_reduce!`, which can accept `Broadcasted` and `StencilBroadcasted` objects as well.

In addition, the `column_integral_definite!` and `column_integral_indefinite!` functions have been redefined in terms of `column_reduce!` and `column_accumulate!`, respectively. This API will generally allow us to define arbitrary "vertical sweep" operations without the need for custom kernel functions or external caches, which should allow us to simplify the non-orographic gravity wave parametrization in ClimaAtmos.

This PR also updates ClimaComms to 0.6.4 in all Manifest files to make use of the new device-flexible `assert` macro.

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

charleskawczynski commented on src/Operators/integrals.jl:
```diff
@@ -1,315 +1,371 @@
+                ϕ_new_implicit_residual,
+                RootSolvers.NewtonsMethodAD(ϕ_prev_implicit),
+                RootSolvers.CompactSolution(),
+            ).root
```
Should we do error handling here? (what if `find_zero` does not converge?)

charleskawczynski commented on src/Operators/integrals.jl:
```diff
@@ -1,315 +1,371 @@
+    Operators.StencilBroadcasted,
+}
+
+extract_column(colidx) = Base.Fix2(extract_column, colidx)
+extract_column(x, _) = x
+extract_column(field::Fields.Field, colidx) = field[colidx]
+extract_column(op::FiniteDifferenceOperator, colidx) =
+    unionall_type(typeof(op))(
+        NamedTuple{keys(op.bcs)}(
+            unrolled_map(extract_column(colidx), values(op.bcs)),
+        ),
+    )
+extract_column(bc::PointwiseOrColumnwiseBroadcasted, colidx) =
+    Base.Broadcast.broadcasted(
+        bc isa Operators.StencilBroadcasted ? extract_column(bc.op, colidx) :
+        bc.f,
+        unrolled_map(extract_column(colidx), bc.args)...,
+    )
```
How is this different from just calling `column`?

charleskawczynski replied:
If this is needed, can we add some unit tests for it?

charleskawczynski commented on src/Operators/integrals.jl:
```diff
@@ -1,315 +1,371 @@
+        reduced_value = column_value_at_level(input, first_level)
+        level_range = (first_level + 1):last_level
+    else
+        reduced_value = init
```
```suggestion
        reduced_value::typeof(init) = init
```
It might turn out to be a good idea to this assertion here, so that the compiler doesn't need to infer through the entire `level_range` (since we're not doing `reduce_value .= ...`)

charleskawczynski replied:
(and same above)

dennisYatunin replied:
There aren't any type instabilities in the GPU tests at the moment, so it doesn't look like the additional typecast makes a difference here.

charleskawczynski reviewed (commented):
Looks good to me, I left some comments, my main comment is: can we add some more granular unit tests?

dennisYatunin commented:
I've enabled GPU support and added more unit tests. Adding the error handling for `find_zero` actually revealed that one of the profiles in ClimaAtmos was not converging for the default absolute tolerance of 0.001, so I changed it to a relative tolerance of 0.001. I'm going to move the device-specific `@assert` to ClimaComms, and then this will be ready to merge in. @charleskawczynski, could you please take another look?
